cpp/src/table/table_device_view.cu

```diff
@@ -86,13 +87,31 @@ extern template bool is_relationally_comparable<table_view>(table_view const& lh
 extern template bool is_relationally_comparable<mutable_table_view>(mutable_table_view const& lhs,
                                                                     mutable_table_view const& rhs);
 
+template <typename TableDeviceView>
+bool is_relationally_comparable(TableDeviceView const& lhs,
+                                TableDeviceView const& rhs,
+                                rmm::cuda_stream_view stream)
+{
+  return thrust::all_of(rmm::exec_policy(stream),
+                        thrust::counting_iterator<size_type>(0),
+                        thrust::counting_iterator<size_type>(lhs.num_columns()),
+                        [lhs, rhs] __device__(auto const i) {
+                          return lhs.column(i).type() == rhs.column(i).type() and
+                                 type_dispatcher(lhs.column(i).type(),
+                                                 is_relationally_comparable_functor{});
+                        });
```

Comment:
Why does this need to run on device? All we're doing is looking at the types of the columns. There shouldn't need to be anything on the device here.

Reply:
The column array in a `table_device_view` is in device memory -- this call `lhs.column(i)` requires accessing device memory.

Reply:
https://github.com/rapidsai/cudf/blob/c1638869116aae2c6dde6024394279a2fb79e685/cpp/include/cudf/table/table_device_view.cuh#L57-L62

Reply:
Oh, this is not good. The fact that it's a template hid that this is what's going on. 

This function should definitely not be launching a kernel.  Which means the function probably shouldn't exist for `table_device_view` and only `table_view`, which would require modifying the `row_comparator` here. 

I guess this fix is necessary and fine for now, but we should work to eliminate needing this ASAP.

Reply:
There seems to be no easy way for the `row_lexicographic_comparator` to check comparability here. If we pass non-device views on the [constructor](https://github.com/rapidsai/cudf/blob/87a2ea485ef9a91fb9df46eb6135f083671147cb/cpp/include/cudf/table/row_operators.cuh#L372-L376) instead then the check can be done efficiently in host code. But the comparator needs device-views for it's [`operator()`](https://github.com/rapidsai/cudf/blob/87a2ea485ef9a91fb9df46eb6135f083671147cb/cpp/include/cudf/table/row_operators.cuh#L397) method. If the `row_lexicographic_comparator` were to create the device-views in the constructor we would need to store unique-ptrs in the member data. This would delete the copy-constructor which is needed to pass the functor to [thrust calls for example](https://github.com/rapidsai/cudf/blob/87a2ea485ef9a91fb9df46eb6135f083671147cb/cpp/src/sort/is_sorted.cu#L50-L59). Storing the unique-ptrs as shared-ptrs in the member data would work but means that the comparator cannot be passed through to lower device functions since the copy-constructor would try to increment the shared-ptr reference counter which is stored in host memory.

I don't think we want to pass both the device and non-device table views to the constructor so my inclination here is to make the caller ensure comparability of the table columns before using the `row_lexicographic_comparator`. This would remove the need for a kernel launch (thrust device call) in the constructor.

@devavret Maybe this can be incorporated in your rework of the comparators?